src/Sparrow/Json/BlittableJsonDocumentBuilder.cs

```diff
@@ -55,12 +66,36 @@ public BlittableJsonDocumentBuilder(JsonOperationContext context, JsonParserStat
             Renew(debugTag, mode);
         }
 
+        private Func<bool> GetReadInternalFunction(IJsonParser reader, UsageMode mode)
+        {
+            // Instead of checking mode and parser type repeatedly in the reading loop,
+            // we perform that check once. We return a specialized generic function
+            // (either ReadInternal<WriteNone, TParser> or ReadInternal<WriteFull, TParser>).
+            // This approach enables better inlining and reduces branching inside Read().
+            return mode switch
+            {
+                UsageMode.None => reader switch
+                {
+                    ObjectJsonParser => ReadInternal<WriteNone, ObjectJsonParser>,
+                    UnmanagedJsonParser => ReadInternal<WriteNone, UnmanagedJsonParser>,
+                    _ => ReadInternal<WriteNone, IJsonParser>
+                },
+                _ => reader switch
+                {
+                    ObjectJsonParser => ReadInternal<WriteFull, ObjectJsonParser>,
+                    UnmanagedJsonParser => ReadInternal<WriteFull, UnmanagedJsonParser>,
+                    _ => ReadInternal<WriteFull, IJsonParser>
+                }
+            };
+        }
+
         public void Reset()
         {
             AssertNotDisposed();
 
             _debugTag = null;
             _mode = UsageMode.None;
+            _readInternalFunc = GetReadInternalFunction(_reader, _mode);
```

Comment:
Shouldn't `Reset()` just set it to `null`?

Reply:
Also shouldn't we set it to `null` on dispose?

Reply:
Yes we could, that's going to be even better (semantics wise).